File: app/src/main/java/com/maxandron/daggerissue325java/bar/BarProvider.java
package com.maxandron.daggerissue325java.bar;

import dagger.Module;
import dagger.Provides;

@Module
public class BarProvider {
    private final BarActivity view;

    public BarProvider(BarActivity view) {
        this.view = view;
    }

    @Provides
    @BarScope
    BarActivity provideView() {
        return view;
    }

    @Provides
    @BarScope
    Bar provideFooOrBar() {
        return new Bar();
    }
}

File: app/src/main/java/com/maxandron/daggerissue325java/bar/BarProvider.kt
package com.maxandron.daggerissue325java.bar

import dagger.Module
import dagger.Provides

@Module
class BarProvider(private val view: BarActivity) {

    @Provides @BarScope fun provideView() = view

    @Provides @BarScope fun provideFooOrBar() = Bar()
}
